code/game/objects/effects/proximity.dm

```diff
@@ -0,0 +1,96 @@
+/datum/proximity_monitor
+	var/atom/host	//the atom we are tracking
+	var/atom/last_host_loc
+	var/list/checkers //list of /obj/effect/abstract/proximity_checkers
+	var/current_range
+	var/ignore_if_not_on_turf	//don't check turfs in range if the host's loc isn't a turf
+
+/datum/proximity_monitor/New(atom/_host, range, _ignore_if_not_on_turf = TRUE, datum/callback/_on_proximity = null)
+	host = _host
+	last_host_loc = _host.loc
+	ignore_if_not_on_turf = _ignore_if_not_on_turf
+	if(istype(host, /atom/movable))	//no need for this on turfs
+		START_PROCESSING(SSobj, src)
+	SetRange(range)
+
+/datum/proximity_monitor/Destroy()
+	host = null
+	QDEL_LIST(checkers)
+	STOP_PROCESSING(SSobj, src)
+	return ..()
+
+/datum/proximity_monitor/process()

```

Comment:
Polling is very inaccurate. you're using SSobj, like, holy shit, that only runs every two seconds. You can move 20 tiles in that time!